app/src/lib/helpers/regex.ts

```diff
@@ -0,0 +1,41 @@
+/**
+ * get all regex captures within a body of text
+ * @param text string to search
+ * @param re regex to search with. must have global option and one capture
+ * @returns set of strings captured by supplied regex
+ */
+export async function getCaptures(
+  text: string,
+  re: RegExp
+): Promise<Set<Array<string>>> {
+  const matches = await getMatches(text, re)
+  const captures = matches.reduce(
+    (captures, match) => captures.add(match.slice(1)),
+    new Set<Array<string>>()
+  )
+  return captures
+}
+
+/**
+ * get all regex matches within a body of text
+ * @param text string to search
+ * @param re regex to search with. must have global option
+ * @returns set of strings captured by supplied regex
+ */
+export async function getMatches(
+  text: string,
+  re: RegExp
+): Promise<Array<RegExpExecArray>> {
+  const matches = new Array<RegExpExecArray>()
+  const getNextMatch = () =>
+    new Promise(resolve => {
+      const match = re.exec(text)
+      if (match !== null) {
+        matches.push(match)
+        resolve(getNextMatch())
```

Comment:
It's unclear to me when this recurrence would end. When would we satisfy the base condition: `match === null`? `const match = re.exec(text)` doesn't seem to ever change since the text being matched never changes.

Could you move the base condition to the top?

```ts
 const getNextMatch = () =>
    new Promise(resolve => {
      const match = re.exec(text)

      if (match === null) {
        resolve(matches)
        return
      }

      matches.push(match)
      resolve(getNextMatch())
  })
```